### PR TITLE
[NVPTX] Make i16x2 a native type and add supported vec instructions

### PR DESCRIPTION
**Pull request by ThomasRaoux**
On sm_90 some instructions now support i16x2 which allows hardware to execute more efficiently add, min and max instructions.

In order to support that we need to make i16x2 a native type in the backend. This does the necessary changes to make i16x2 a native type and adds support for the instructions natively supporting i16x2.

This caused a negative test in nvptx slp to start passing. Changed the test to a positive one as the IR is correctly vectorized.

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 1]
```diff
@@ -1262,10 +1264,11 @@ bool NVPTXDAGToDAGISel::tryLDGLDU(SDNode *N) {
     EltVT = EltVT.getVectorElementType();
     // vectors of f16 are loaded/stored as multiples of v2f16 elements.
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1265]: Comment needs updating/generalizing to 2-element vectors of 16-bit types.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp [hunk 2]
```diff
@@ -1678,7 +1681,8 @@ bool NVPTXDAGToDAGISel::tryStore(SDNode *N) {
   MVT ScalarVT = SimpleVT.getScalarType();
   unsigned toTypeWidth = ScalarVT.getSizeInBits();
   if (SimpleVT.isVector()) {
-    assert((StoreVT == MVT::v2f16 || StoreVT == MVT::v2bf16) &&
+    assert((StoreVT == MVT::v2f16 || StoreVT == MVT::v2bf16 ||
+            StoreVT == MVT::v2i16) &&
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1685]: We may want to extract it into an `isv2x16vt` helper function, so we don't clutter code with this boilerplate comparison in many places.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 3]
```diff
@@ -149,12 +150,13 @@ static bool IsPTXVectorType(MVT VT) {
   }
 }
 
-static bool Isv2f16Orv2bf16Type(EVT VT) {
-  return (VT == MVT::v2f16 || VT == MVT::v2bf16);
+static bool Isv2f16Orv2bf16Orv2i16Type(EVT VT) {
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:153]: Ah, we already have that helper. Move it upwards and use it for the comparisons above.

Maybe rename it to something less cumbersome. `Isv2x16Type` ?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 4]
```diff
@@ -207,8 +209,13 @@ static void ComputePTXValueVTs(const TargetLowering &TLI, const DataLayout &DL,
       // Vectors with an even number of f16 elements will be passed to
       // us as an array of v2f16/v2bf16 elements. We must match this so we
       // stay in sync with Ins/Outs.
-      if ((Isf16Orbf16Type(EltVT.getSimpleVT())) && NumElts % 2 == 0) {
-        EltVT = EltVT == MVT::f16 ? MVT::v2f16 : MVT::v2bf16;
+      if ((Is16bitsType(EltVT.getSimpleVT())) && NumElts % 2 == 0) {
+        if (EltVT == MVT::f16)
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:213]: Nit: I'd use `switch(EltVT)`. This is fine, too.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 5]
```diff
@@ -2122,7 +2192,8 @@ SDValue NVPTXTargetLowering::LowerEXTRACT_VECTOR_ELT(SDValue Op,
   // Extract individual elements and select one of them.
   SDValue Vector = Op->getOperand(0);
   EVT VectorVT = Vector.getValueType();
-  assert(VectorVT == MVT::v2f16 && "Unexpected vector type.");
+  assert((VectorVT == MVT::v2f16 || VectorVT == MVT::v2i16) &&
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2195]: We may be missing a v2bf16 here.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 6]
```diff
@@ -2470,7 +2541,7 @@ SDValue NVPTXTargetLowering::LowerLOAD(SDValue Op, SelectionDAG &DAG) const {
 
   // v2f16 is legal, so we can't rely on legalizer to handle unaligned
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2574]: Comment needs updating.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 7]
```diff
@@ -2515,13 +2586,13 @@ SDValue NVPTXTargetLowering::LowerSTORE(SDValue Op, SelectionDAG &DAG) const {
...
-  if (Isv2f16Orv2bf16Type(VT) &&
+  if (Isv2f16Orv2bf16Orv2i16Type(VT) &&
       !allowsMemoryAccessForAlignment(*DAG.getContext(), DAG.getDataLayout(),
                                       VT, *Store->getMemOperand()))
     return expandUnalignedStore(Store, DAG);
 
-  // v2f16 and v2bf16 don't need special handling.
-  if (VT == MVT::v2f16 || VT == MVT::v2bf16)
+  // v2f16, v2bf16 and v2i16 don't need special handling.
+  if (VT == MVT::v2f16 || VT == MVT::v2bf16 || VT == MVT::v2i16)
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2627]: Use Isv2f16Orv2bf16Orv2i16Type() ?

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td [hunk 8]
```diff
@@ -747,6 +754,13 @@ def SELP_f16x2rr :
               [(set Int32Regs:$dst,
                     (select Int1Regs:$p, (v2f16 Int32Regs:$a), (v2f16 Int32Regs:$b)))]>;
 
+def SELP_i16x2rr :
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:757]: Needs sm90/ptx80 predicates?

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+  ret i16 %e
+}
+
+; CHECK-LABEL: test_add(
+; CHECK-DAG:  ld.param.u32    [[A:%r[0-9]+]], [test_add_param_0];
+; CHECK-DAG:  ld.param.u32    [[B:%r[0-9]+]], [test_add_param_1];
+;
+; CHECK-I16x2-NEXT:  add.s16x2   [[R:%r[0-9]+]], [[A]], [[B]];
+;
+;	CHECK-NOI16x2-DAG: mov.b32 	{[[RS0:%rs[0-9]+]], [[RS1:%rs[0-9]+]]}, [[A]];
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/i16x2-instructions.ll:72]: Is indentation here intentional? The check prefix itself provides sufficient clue w/o extra spaces.

Speaking of the CHECK prefixes. I'd rename them:
CHECK->COMMON
CHECK-I16x2 -> I16x2
CHECK-NOI16x2 -> NO-I16x2

**Review by Artem-B — COMMENTED** [review]
Thank you for the patch.
It looks fine in general. One thing I'd suggest is to extract generic tablegen changes (adding specific parameter types) into a separate patch so they do not clutter the v2i16 changes.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Done, moved the function in the helper to be able to use it in this file.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 3 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: done.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:209]: EltVT is not an enum so using it in a switch doesn't work out of the box. So I left it like that, let me know if you see a better way.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 5 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Moved to using Isv2x16Type so that addresses it.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 7 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: done.

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll (hunk 9 above)
Comment (ThomasRaoux) [reply llvm/test/CodeGen/NVPTX/i16x2-instructions.ll]: Removed the indentation and changed the prefix names

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 8 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:757]: no this is supported on all platforms as it becomes just a int32 select.

**Review by ThomasRaoux — COMMENTED** [review]
Thanks, I rebased on top of the tablegen changes and addressed the comments.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 8 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:757]: In that case, can we just use the existing `SELP_PATTERN`?
While we're at that, do the same for fp16/bf16.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:209]: `switch(EltVT.SimpleTy)` should work.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 10]
```diff
@@ -584,6 +617,22 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
...
+  setI16x2OperationAction(ISD::SMIN, MVT::v2i16, Legal, Custom);
+  setI16x2OperationAction(ISD::SMAX, MVT::v2i16, Legal, Custom);
+  setI16x2OperationAction(ISD::UMIN, MVT::v2i16, Legal, Custom);
+  setI16x2OperationAction(ISD::UMAX, MVT::v2i16, Legal, Custom);
+  setI16x2OperationAction(ISD::CTPOP, MVT::v2i16, Legal, Expand);
+  setI16x2OperationAction(ISD::CTLZ, MVT::v2i16, Legal, Expand);
+
+  setI16x2OperationAction(ISD::ADD, MVT::v2i16, Legal, Custom);
+  setI16x2OperationAction(ISD::SUB, MVT::v2i16, Legal, Custom);
+  setI16x2OperationAction(ISD::AND, MVT::v2i16, Legal, Custom);
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:630]: and/or/xor would also boil down to i32 logical ops and probably do not require new GPUs. 
It looks like ADD ended up here by mistake as none of the other logical ops made it to this list.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 8 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Good point, changed it to a pattern for all the cases

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 11]
```diff
@@ -5198,6 +5296,7 @@ static void ReplaceLoadVector(SDNode *N, SelectionDAG &DAG,
   case MVT::v4f16:
   case MVT::v4f32:
   case MVT::v8f16: // <4 x f16x2>
+  case MVT::v8i16: // <4 x i16x2>
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5299]: Looks like we're missing bf16 here.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td [hunk 12]
```diff
@@ -3300,6 +3328,13 @@ def : Pat<(bf16 (extractelt (v2bf16 Int32Regs:$src), 1)),
 def : Pat<(v2bf16 (build_vector (bf16 Int16Regs:$a), (bf16 Int16Regs:$b))),
           (V2I16toI32 Int16Regs:$a, Int16Regs:$b)>;
 
+def : Pat<(i16 (extractelt (v2i16 Int32Regs:$src), 0)),
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3324]: We're doing exactly the same things for f16/bf16. Perhaps we can just coalesce them all into a loop iterating over the types.

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+  ret i16 %e
+}
+
+; COMMON-LABEL: test_add(
+; COMMON-DAG:  ld.param.u32    [[A:%r[0-9]+]], [test_add_param_0];
+; COMMON-DAG:  ld.param.u32    [[B:%r[0-9]+]], [test_add_param_1];
+;
+; I16x2-NEXT:  add.s16x2   [[R:%r[0-9]+]], [[A]], [[B]];
+;
+;	NO-I16x2-DAG: mov.b32 	{[[RS0:%rs[0-9]+]], [[RS1:%rs[0-9]+]]}, [[A]];
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/i16x2-instructions.ll:72]: Indentations are still here.

File: llvm/test/CodeGen/NVPTX/vec-param-load.ll [hunk 14]
```diff
@@ -70,14 +70,10 @@ define <8 x i64> @test_v8i64(<8 x i64> %a) {
...
 ; CHECK-LABEL: test_v16i16(
-; CHECK-DAG: ld.param.v4.u16     {[[V_12_15:(%rs[0-9]+[, ]*){4}]]}, [test_v16i16_param_0+24];
-; CHECK-DAG: ld.param.v4.u16     {[[V_8_11:(%rs[0-9]+[, ]*){4}]]}, [test_v16i16_param_0+16];
-; CHECK-DAG: ld.param.v4.u16     {[[V_4_7:(%rs[0-9]+[, ]*){4}]]}, [test_v16i16_param_0+8];
-; CHECK-DAG: ld.param.v4.u16     {[[V_0_3:(%rs[0-9]+[, ]*){4}]]}, [test_v16i16_param_0];
-; CHECK-DAG: st.param.v4.b16     [func_retval0+0], {[[V_0_3]]}
-; CHECK-DAG: st.param.v4.b16     [func_retval0+8], {[[V_4_7]]}
-; CHECK-DAG: st.param.v4.b16     [func_retval0+16], {[[V_8_11]]}
-; CHECK-DAG: st.param.v4.b16     [func_retval0+24], {[[V_12_15]]}
+; CHECK-DAG: ld.param.v4.u32     {[[V_8_15:(%r[0-9]+[, ]*){4}]]}, [test_v16i16_param_0+16];
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/vec-param-load.ll:73]: Nice. This will make some XLA users happy.

File: llvm/test/Transforms/SLPVectorizer/NVPTX/non-vectorizable-intrinsic-inseltpoison.ll
Comment (Artem-B) [inline llvm/test/Transforms/SLPVectorizer/NVPTX/non-vectorizable-intrinsic-inseltpoison.ll:1]: What's the story with these test files?  Are they relevant to v2i16 changes in the rest of the patch? If not, they should go into a separate patch.

Comment (ThomasRaoux) [reply llvm/test/Transforms/SLPVectorizer/NVPTX/non-vectorizable-intrinsic-inseltpoison.ll:1]: It is linked as making v2i16 a legal type makes the vectorization pass apply vectorization to this IR. I'm not sure why we needed those negative tests in a first place but I don't see a way to separate this out

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 10 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Right, removed the `AND`

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 12 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3324]: you mean just for this pattern or more in general?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 6 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 11 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Added it.

File: llvm/test/CodeGen/NVPTX/i16x2-instructions.ll (hunk 13 above)
Comment (ThomasRaoux) [reply llvm/test/CodeGen/NVPTX/i16x2-instructions.ll]: Oops, somehow might editor was misleading me. Should be good now.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 12 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3324]: Specifically for build_vector and extractelt 0/1 for those types. If you see more opportunities, feel free to coalesce them tool. Now that we've switched to i32 registers for all those types, fair number of operations becomes common for those types. I've tried to merge them, but apparently missed some.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 15]
```diff
@@ -2347,7 +2413,26 @@ SDValue NVPTXTargetLowering::LowerFROUND64(SDValue Op,
...
+  SmallVector<SDValue> VecElements;
+  for (int I = 0, E = Op.getValueType().getVectorNumElements(); I < E; I++) {
+    SmallVector<SDValue> ScalarArgs;
+    for (int J = 0, NumOp = Op.getNumOperands(); J < NumOp; J++) {
+      SDValue Ext =
+          DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, EltVT, Op->getOperand(J),
+                      DAG.getIntPtrConstant(I, DL));
+      ScalarArgs.push_back(Ext);
+    }
+    VecElements.push_back(DAG.getNode(Op.getOpcode(), DL, EltVT, ScalarArgs));
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2436]: Nit. I'm very tempted to suggest coalescing the loop into something like this:
```
llvm::transform(Op.ops(), back_inserter(ScalarArgs), [&](auto O){
  return DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, EltVT, O->getOperand(J),  DAG.getIntPtrConstant(I, DL))
})
```
That may be pushing it too far, but the inner per-operand loop itself does look like a prime candidate for replacement  with a `range for` as we do not use `J` for anything other than getting the operand.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 12 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3322]: We still need to specialize based on the scalar type. Is there a way to write a foreach loop that handle this case?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 11 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: I do not see it in the diff.

Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: It's there now, I believe there was some lag in GitHub UI

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 12 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3322]: At the moment it's a bit cumbersome. You iterate with `foreach foo =  [a,b,c,d] ` over a scalar value. Could be a string with the register type.
Then create an instance of a helper class specialized by it, and which will have fields populated with the values you will need. Somewhat extreme example is WMMA_REGS in llvm/include/llvm/IR/IntrinsicsNVVM.td

We have a pending patch which will add lambdas and functions to tablegen which would make all of this easier, but for now it's awkward. If it turns out to be too much hassle, just place f16/bf16/i16 extractelt/build_vector next to each other, so it's easy to see that they are repeated.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 11 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Yup. I see the changes now.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td (hunk 12 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: ok I coalesced the extract case where the scalar type can be inferred and move the build_vector together.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 15 above)
Comment (ThomasRaoux) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: sure, changed it.

**Review by Artem-B — APPROVED** [review]
LGTM. Thank you!

**Comment by Artem-B** [comment]
Looks like we've missed lowering of bitcasts between v2f16 and v2i16 and it breaks XLA.

```
LLVM ERROR: Cannot select: t119: v2f16 = bitcast t118
  t118: v2i16 = or t375, t401
    t375: v2i16 = BUILD_VECTOR t374, t372
      t374: i16 = select t247, Constant:i16<8960>, t360
```

**Comment by ThomasRaoux** [comment]
> Looks like we've missed lowering of bitcasts between v2f16 and v2i16 and it breaks XLA.
> 
> ```
> LLVM ERROR: Cannot select: t119: v2f16 = bitcast t118
>   t118: v2i16 = or t375, t401
>     t375: v2i16 = BUILD_VECTOR t374, t372
>       t374: i16 = select t247, Constant:i16<8960>, t360
> ```

Oops, I can send a patch today unless someone else already has a fix.

**Comment by gribozavr** [comment]
Thank you @Artem-B for the breakage analysis!

@ThomasRaoux I hope you don't mind, I'm going to revert this change to unbreak XLA.

**Comment by ThomasRaoux** [comment]
> Thank you @Artem-B for the breakage analysis!
> 
> @ThomasRaoux I hope you don't mind, I'm going to revert this change to unbreak XLA.

Can we fix it forward? I'll prepare a patch.

**Comment by Artem-B** [comment]
We'll probably also need similar bitcast lowering for v2bf16. I suspect now that v2i16 are available we'll see LLVM picking them for various loads/stores and bf16 will likely end up using that path, too. Not sure if we'll need bitcasts between v2f16 and v2bf16 in practice, but it would not hurt to add them just in case, too.

**Comment by ThomasRaoux** [comment]
> We'll probably also need similar bitcast lowering for v2bf16. I suspect now that v2i16 are available we'll see LLVM picking them for various loads/stores and bf16 will likely end up using that path, too. Not sure if we'll need bitcasts between v2f16 and v2bf16 in practice, but it would not hurt to add them just in case, too.

sure, let me add those.

**Comment by Artem-B** [comment]
> Can we fix it forward? I'll prepare a patch.

The fix itself should be simple (tablegen pattern converting bitcast to a proxyreg?) but we'll also need tests. Should be doable, but I don't know @gribozavr 's time constraints.

**Comment by Artem-B** [comment]
In any case, it probably does not matter much one way or another -- it's just one more cherry-picked revision to include into a new pull request.

**Comment by ThomasRaoux** [comment]
looking at it we do have some tests for bitcast 2xhalf -> 2xi16:
https://github.com/llvm/llvm-project/blob/f2b04434fc49e41444d1a1a4065e8fbe249538c2/llvm/test/CodeGen/NVPTX/f16x2-instructions.ll#L998

**Comment by Artem-B** [comment]
Hmm. Why did they work then? Loads/stores/bitcasts for v2i16 should not have dependent on whether the target is sm_90 or not. Or maybe it did, if we set lowering action to Expand on older GPUs. 

You may try running the tests with `-mcpu=sm_90`

**Comment by ThomasRaoux** [comment]
> Hmm. Why did they work then? Loads/stores/bitcasts for v2i16 should not have dependent on whether the target is sm_90 or not. Or maybe it did, if we set lowering action to Expand on older GPUs.
> 
> You may try running the tests with `-mcpu=sm_90`

I tried with `-mcpu=sm_90 -mattr=+ptx80` and without it and I can't reproduce the problem.

**Comment by Artem-B** [comment]
Another theory is that the bitcast in the tests didn't actually make it to the lowering and we only had to deal with loads/stores. 
https://godbolt.org/z/81e31jj8a
nvptx-isel appears to convert IR straight into
```
# *** IR Dump After NVPTX DAG->DAG Pattern Instruction Selection (nvptx-isel) ***:
# Machine code for function test_bitcast_2xhalf_to_2xi16: IsSSA, TracksLiveness

bb.0 (%ir-block.0):
  %0:int32regs = LD_i32_avar 0, 4, 1, 0, 32, &test_bitcast_2xhalf_to_2xi16_param_0 :: (dereferenceable invariant load (s32) from `ptr addrspace(101) null`, addrspace 101)
  StoreRetvalI32 killed %0:int32regs, 0 :: (store (s32), align 1)
  Return
```

We may need something more elaborate, like manually constructing v2i16 from an input i16.

**Comment by gribozavr** [comment]
> Can we fix it forward? I'll prepare a patch.

The HEAD has been broken for more than a day, and we need to get to green ASAP. I pushed the revert.

**Comment by Artem-B** [comment]
Aha. This reproduces the problem: https://godbolt.org/z/K1c4PqYoP

**Comment by ThomasRaoux** [comment]
> Aha. This reproduces the problem: https://godbolt.org/z/K1c4PqYoP

Thanks for the repro. I'm going to land it again with the fix (https://github.com/llvm/llvm-project/pull/65799/commits/2bd3ce7e092a8afa0718dc6a43d53fa3b180d292)

**Comment by cheshire** [comment]
This causes a nearly ~2x regression in many Triton kernels on Ampere, we'll post a reproducer.
